core/src/avm2/globals/string.rs

```diff
@@ -120,6 +120,243 @@ fn char_code_at<'gc>(
+            return Ok((-1).into());
+        }
+    }
+
+    Ok(Value::Undefined)
+}
+
+/// Implements `String.lastIndexOf`
+fn last_index_of<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let Some(this) = this {
+        let this = Value::from(this)
+            .coerce_to_string(activation)?
+            .encode_utf16()
+            .collect::<Vec<u16>>();
+        let pattern = match args.get(0) {
+            None => return Ok(Value::Undefined),
+            Some(s) => s
+                .clone()
+                .coerce_to_string(activation)?
+                .encode_utf16()
+                .collect::<Vec<_>>(),
+        };
+        let start_index = match args.get(1) {
+            None | Some(Value::Undefined) => this.len(),
+            Some(n) => n.coerce_to_i32(activation)?.max(0) as usize,
+        };
+
+        return if pattern.is_empty() {
+            // Empty pattern is found immediately.
+            Ok(start_index.into())
+        } else if let Some((i, _)) = this[..]
+            .windows(pattern.len())
+            .enumerate()
+            .take(start_index + 1)
+            .rev()
+            .find(|(_, w)| *w == &pattern[..])
+        {
+            Ok(i.into())
+        } else {
+            // Not found
+            Ok((-1).into())
+        };
+    }
+
+    Ok(Value::Undefined)
+}
+
+/// Implements `String.match`
+fn match_s<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    this: Option<Object<'gc>>,
+    args: &[Value<'gc>],
+) -> Result<Value<'gc>, Error> {
+    if let (Some(this), pattern) = (this, args.get(0).unwrap_or(&Value::Undefined)) {
+        let this = Value::from(this).coerce_to_string(activation)?;
+
+        let regexp_class = activation.avm2().classes().regexp;
+        let pattern = if !pattern.is_of_type(activation, regexp_class)? {
+            let string = pattern.coerce_to_string(activation)?;
+            regexp_class.construct(activation, &[Value::String(string)])?
+        } else {
+            pattern.coerce_to_object(activation)?
+        };
+
+        if let Some(mut regexp) = pattern.as_regexp_mut(activation.context.gc_context) {
+            let mut storage = ArrayStorage::new(0);
+            if regexp.global() {
+                let mut last = regexp.last_index();
+                regexp.set_last_index(0);
+                while let Some(result) = regexp.exec(&this) {
+                    if regexp.last_index() == last {
+                        break;
+                    }
+                    storage.push(
+                        AvmString::new(
+                            activation.context.gc_context,
+                            this[result.range()].to_string(),
+                        )
+                        .into(),
+                    );
+                    last = regexp.last_index();
+                }
+                regexp.set_last_index(0);
```

Comment:
If possible, let's try to recreate the weird Flash behavior of `lastIndex` here following https://github.com/adobe/avmplus/blob/master/core/RegExpObject.cpp#L492-L495

You can add a comment that this behavior doesn't match what the API docs claim.

Reply:
at first I thought that [here](https://github.com/adobe/avmplus/blob/master/core/RegExpObject.cpp#L485) once loop is about to finish `m_lastIndex` may not always be set to `0`, but after looking at [_exec](https://github.com/adobe/avmplus/blob/master/core/RegExpObject.cpp#L377) from avmplus I think that 
```
                regexp.set_last_index(0);
                if old_last_index == regexp.last_index() {
                    regexp.set_last_index(1);
                }
```
will work just like `flashplayer`